File: Project/Generic_DS_Lib/DoublyCircularLL.java
import java.lang.*;
import java.util.*;

class node
{
    public int data;
    public node next;       // struct node * next;
    public node prev;
}

class DoublyCLL
{
    public node Head;
    public node Tail;
    public int Count;

    public DoublyCLL()
    {
        Head = null;
        Tail = null;
        Count = 0;
    }

    public void InsertFirst(int no)
    {
        node newn = new node();

        newn.data = no;
        newn.next = null;
        newn.prev = null;

        if(Count == 0)
        {
            Head = newn;
            Tail = newn;
        }
        else
        {
            newn.next = Head;            
            Head.prev = newn;
            Head = newn;
        }
        Head.prev = Tail;
        Tail.next = Head;
        Count++;
    }
    
    public void InsertLast(int no)
    {
        node newn = new node();

        newn.data = no;
        newn.next = null;
        newn.prev = null;

        if(Count == 0) 
        {
            Head = newn;
            Tail = newn;
        }
        else
        {
            Tail.next = newn;
            newn.prev = Tail;
            Tail = newn;            
        }
        Head.prev = Tail;
        Tail.next = Head;
        Count++;
    }

    public void Display()
    {
        node temp = Head;

        if(Count == 0)
        {
            System.out.println("Link list is empty");
        }
        else
        {
            for(int iCnt = 1; iCnt<= Count; iCnt++)
            {
                System.out.print("|"+temp.data+"|<=>");
                temp = temp.next;
            }          
            System.out.println("");  
        }
    }

    public int CountNode()
    {
        return Count;
    }
    
    public void DeleteFirst()
    {
        if(Count == 0)
        {
            return;
        }
        else if(Count == 1)
        {
            Head = null;
            Tail = null;
        }
        else
        {
            node temp = Head;
            Head = Head.next;
            Tail.next = Head;
            temp = null;
        }
        Count--;
    }

    public void DeleteLast()
    {
        if(Count == 0)
        {
            return;
        }
        else if(Count == 1)
        {
            Head = null;
            Tail = null;
        }
        else
        {
            node temp = Tail;
            Tail = Tail.prev;
            temp = null;
        }
        Tail.next = Head;
        Head.prev = Tail;
        Count--;
    }
    
    public void InsertAtPos(int no, int pos)
    {
        if((pos < 1) || (pos > Count+1))
        {
            System.out.println("Invalid Position");
            return;
        }

        if(pos == 1)
        {
            InsertFirst(no);
        }
        else if(pos == Count+1)
        {
            InsertLast(no);
        }
        else
        {
            node newn = new node();
            newn.data = no;
            newn.next = null;
            newn.prev = null;

            node temp = Head;

            for(int i = 1; i< pos-1; i++)
            {
                temp = temp.next;
            }

            newn.next = temp.next;
            temp.next.prev = newn;
            temp.next = newn;
            newn.prev = temp;

            Count++;
        }
    }
    
    public void DeleteAtPos(int pos)
    {
        if((pos < 1) || (pos > Count))
        {
            System.out.println("Invalid Position");
            return;
        }

        if(pos == 1)
        {
            DeleteFirst();
        }
        else if(pos == Count)
        {
            DeleteLast();
        }
        else
        {
            node temp = Head;

            for(int i = 1; i< pos-1; i++)
            {
                temp = temp.next;
            }

            temp.next = temp.next.next;
            temp.next.prev = null;
            temp.next.prev = temp;

            Count--;
        }
    }
}

class DoublyCircularLL
{
    public static void main(String arg[])
    {
        DoublyCLL obj = new DoublyCLL();

        obj.InsertFirst(51);
        obj.InsertFirst(21);
        obj.InsertFirst(11);
        obj.Display();
        System.out.println("Number of nodes are : "+obj.CountNode());

        obj.InsertLast(101);
        obj.InsertLast(111);
        obj.Display();
        System.out.println("Number of nodes are : "+obj.CountNode());

        obj.InsertAtPos(55,4);
        obj.Display();
        System.out.println("Number of nodes are : "+obj.CountNode());

        obj.DeleteAtPos(4);
        obj.Display();
        System.out.println("Number of nodes are : "+obj.CountNode());
        
        obj.DeleteFirst();
        obj.Display();
        System.out.println("Number of nodes are : "+obj.CountNode());

        obj.DeleteLast();
        obj.Display();
        System.out.println("Number of nodes are : "+obj.CountNode());
    }
}

File: Project/Generic_DS_Lib/DoublyCircularLL.cpp
#include <iostream>
using namespace std;

struct node
{
    int data;
    struct node *next;
    struct node *prev;
};

typedef struct node NODE;
typedef struct node* PNODE;

class DoublyCLL
{
    private:
        PNODE head;
        PNODE tail;
        int CountNode;

    public:
        DoublyCLL();

        void InsertFirst(int);
        void InsertLast(int);
        void InsertAtPos(int, int);

        void DeleteFirst();
        void DeleteLast();
        void DeleteAtPos(int);

        int Count();
        void Display();

};

DoublyCLL::DoublyCLL()
{
    head = NULL;
    tail = NULL;
    CountNode = 0;
}

void DoublyCLL::InsertFirst(int iNo)
{
    PNODE newn = NULL;

    newn = new NODE;

    newn->next = NULL;
    newn->prev = NULL;
    newn->data = iNo;

    if(CountNode == 0)
    {
        head = newn;
        tail = newn;
    }
    else
    {
        newn->next = head;
        head->prev = newn;
        head = newn;
        head->prev = tail;
    }

    head->prev = tail;
    tail->next = head;
    CountNode++;
}

void DoublyCLL::InsertLast(int iNo)
{
    PNODE newn = NULL;
    newn = new NODE;

    newn->data = iNo;
    newn->next = NULL;
    newn->prev = NULL;

    if(CountNode == 0)  // if((Head == NULL) && (Tail == NULL))
    {
        head = newn;
        tail = newn;
    }
    else
    {
           tail->next = newn;
           newn->prev = tail;

           tail = newn;
    }

    head->prev = tail;
    tail->next = head;

    CountNode++;
}

void DoublyCLL::InsertAtPos(int iNo, int iPos)
{
    if((iPos < 1) || (iPos > CountNode+1))
        {
            cout<<"Invalid position\n";
            return;
        }

        if(iPos == 1)
        {
            InsertFirst(iNo);
        }
        else if(iPos == CountNode+1)
        {
            InsertLast(iNo);
        }
        else
        {
            PNODE newn = NULL;
            newn = new NODE;

            newn->data = iNo;
            newn->next = NULL;
            newn->prev = NULL;

            PNODE temp = head;

            for(int i = 1; i<iPos-1; i++)
            {
                temp = temp->next;
            }

            newn->next = temp->next;
            temp->next->prev = newn;
            temp->next = newn;
            newn->prev = temp;

            CountNode++;
        }
}

void DoublyCLL::DeleteFirst()
{
    if(CountNode == 0)
    {
            return;
    }
    else if(CountNode == 1)
    {
            delete head;
            head = NULL;
            tail = NULL;
    }
    else
    {
            head = head->next;
            delete tail->next;          // delete Head->prev;
    }

    tail->next = head;
    head->prev = tail;
    CountNode--;
}

void DoublyCLL::DeleteLast()
{
    if(CountNode == 0)
    {
            return;
    }
    else if(CountNode == 1)
    {
            delete tail;
            head = NULL;
            tail = NULL;
    }
    else
    {
            tail = tail->prev;
            delete tail->next;          // delete Head->prev;
    }

    tail->next = head;
    head->prev = tail;
    CountNode--;
}

void DoublyCLL::DeleteAtPos(int iPos)
{
    if((iPos < 1) || (iPos > CountNode))
        {
            cout<<"Invalid position\n";
            return;
        }

        if(iPos == 1)
        {
            DeleteFirst();
        }
        else if(iPos == CountNode)
        {
            DeleteLast();
        }
        else
        {
             PNODE temp = head;

            for(int i = 1; i<iPos-1; i++)
            {
                temp = temp->next;
            }

            temp->next = temp->next->next;
            delete temp->next->prev;
            temp->next->prev = temp;

            CountNode--;
        }
}

void DoublyCLL::Display()
{
    int iCnt = 0;
       PNODE temp = head;

       if(CountNode == 0)
       {
            cout<<"Link list is empty";
       }
       else
       {
            for(iCnt = 1; iCnt<= CountNode; iCnt++)
            {
                    cout<<"|"<<temp->data<<"|<=>";
                    temp = temp->next;
            }
       }
       cout<<endl;
}

int DoublyCLL::Count()
{
    return CountNode;
}

int main()
{
    DoublyCLL obj;
    int iRet = 0;

    obj.Display();
    obj.InsertFirst(21);
    obj.InsertFirst(11);
    obj.Display();
    iRet = obj.Count();
    cout<<"Number of nodes are : "<<iRet<<endl;

    obj.InsertLast(31);
    obj.InsertLast(41);
    obj.Display();
    iRet = obj.Count();
    cout<<"Number of nodes are : "<<iRet<<endl;

    obj.DeleteFirst();
    obj.Display();
    iRet = obj.Count();
    cout<<"Number of nodes are : "<<iRet<<endl;

    obj.DeleteLast();
    obj.Display();
    iRet = obj.Count();
    cout<<"Number of nodes are : "<<iRet<<endl;

    obj.InsertAtPos(25,2);
    obj.Display();
    iRet = obj.Count();
    cout<<"Number of nodes are : "<<iRet<<endl;

    obj.DeleteAtPos(2);
    obj.Display();
    iRet = obj.Count();
    cout<<"Number of nodes are : "<<iRet<<endl;

    return 0;
}